LS-455-gke-nav-and-doc (#37)
datastax/pulsar-docs: modules/ROOT/pages/install-gke.adoc

= Luna Streaming Installation on Google Kubernetes Environment

> This document will cover installation of Luna Streaming on the Google Kubernetes Environment, or GKE. 

Reviewer: ```suggestion
This document will cover installation of Luna Streaming on the Google Kubernetes Environment (GKE). 
```

For an example set of production cluster values, see the DataStax production-ready https://github.com/datastax/pulsar-helm-chart[Helm chart]. +

DataStax recommends these hardware resources for running Luna Streaming in a Kubernetes environment: +

include::partial$helm-chart-recommendations.adoc[]

For the local machine running the Helm chart, you will need: +

* https://helm.sh/docs/[Helm] version 3 +

> * A Google account with priveleges to install and configure Kubernetes resources. 

Reviewer: ```suggestion
* A Google account with privileges to install and configure Kubernetes resources. 
```

== Helm chart `values.yaml`

> The https://github.com/datastax/pulsar-helm-chart/blob/master/helm-chart-sources/pulsar/values.yaml[values.yaml] will deploy Pulsar with:

Reviewer: ```suggestion
The https://github.com/datastax/pulsar-helm-chart/blob/master/helm-chart-sources/pulsar/values.yaml[`values.yaml`] will deploy Pulsar with:
```

> * 1 ZooKeeper cluster with `replicaCount:3` +
> * 1 Bookkeeper cluster with `replicaCount:3` +
> * 1 Broker cluster with `replicaCount: 3` +
> ** Ledgers: `defaultEnsembleSize: 2`, `defaultAckQuorum: 2`, `defaultWriteQuorum: 2` +
> * 1 Function cluster with a `replicaCount: 2` +

Reviewer: ```suggestion
* 1 ZooKeeper cluster with `replicaCount:3`
* 1 Bookkeeper cluster with `replicaCount:3`
* 1 Broker cluster with `replicaCount: 3`
** Ledgers: `defaultEnsembleSize: 2`, `defaultAckQuorum: 2`, `defaultWriteQuorum: 2`
* 1 Function cluster with a `replicaCount: 2`
```

=== `default_storage`

> The `default_storage` parameter in `values.yaml` controls the default storage class for all persistent volumes created by the Helm chart. +

Reviewer: ```suggestion
The `default_storage` parameter in `values.yaml` controls the default storage class for all persistent volumes created by the Helm chart.
```

[source,yaml]
----
default_storage:
  existingStorageClassName: default
----

> For a component like BookKeeper, which requires *stateful* storage, we need to override the `default_storage` class when the BookKeeper Persistent Volume Claims (PVCs) are created. +

Reviewer: ```suggestion
For a component like BookKeeper, which requires stateful storage, we need to override the `default_storage` class when the BookKeeper Persistent Volume Claims (PVCs) are created.
```

> There are two ways to override `default_storage`: +

Reviewer: ```suggestion
There are two ways to override `default_storage`:
```

> 1. Leave `existingStorageClassName` blank and specify the storage class parameters below. +

Reviewer: ```suggestion
* Leave `existingStorageClassName` blank and specify the storage class parameters below.
```

+
[source,yaml]
----
  volumes:
    journal:
      name: journal
      size: 20Gi
        existingStorageClassName: 
        storageClass:
         provisioner: kubernetes.io/gce-pd
         type: pd-ssd
         fsType: ext4
         extraParams:
            replication-type: none
----

> 2. Create a custom storage configuration as a `yaml` file (https://github.com/datastax/pulsar-helm-chart/blob/master/helm-chart-sources/pulsar/templates/bookkeeper/bookkeeper-storageclass.yaml[like this one]) and tell the Helm chart to use that storage configuration when it creates the BookKeeper PVCs. +

Reviewer: ```suggestion
* Create a custom storage configuration as a `yaml` file (https://github.com/datastax/pulsar-helm-chart/blob/master/helm-chart-sources/pulsar/templates/bookkeeper/bookkeeper-storageclass.yaml[like the DataStax example]) and tell the Helm chart to use that storage configuration when it creates the BookKeeper PVCs.
```

+
[source,yaml]
----
  volumes:
    journal:
      name: journal
      size: 20Gi
        existingStorageClassName: bookkeeper-storageclass.yaml
----

=== GKE-specific storage class 

> Here is an example `storage-values.yaml` as used above for the BookKeeper node. Paste it into a `yaml` file to provision a persistent SSD with the ext4 file system.

Reviewer: ```suggestion
Here is an example `storage-values.yaml` as used above for the BookKeeper node. Paste it into a `yaml` file to provision a persistent SSD with the `ext4` file system.
```

[source,yaml]
----
storage_class:
    provisioner: kubernetes.io/gce-pd
    type: pd-ssd
    fsType: ext4
    extraParams:
        replication-type: none
----

== Install Luna Streaming to GKE

Once you have added your storage settings to the `storage-values.yaml` file, install the Helm chart. 

. Create the namespace; in this example, we use `pulsar`.
+
----
kubectl create namespace pulsar
----

. Run this Helm command using the `storage-values.yaml` provisioning:
+
----
helm install pulsar datastax-pulsar/pulsar --namespace pulsar --values storage_values.yaml --create-namespace
----

. To avoid having to specify the `pulsar` namespace on each subsequent command, set the namespace context. Example:
+
----
kubectl config set-context $(kubectl config current-context) --namespace=pulsar
----

. Execute `kubectl get pods -A` to view your running pods. 

> . Once your pods are up and running, continue to xref:quickstart-helm-installs.adoc#manage-pulsar-cluster[Accessing Pulsar Cluster in Cloud] to access and manage your cluster. 
> 
> == Next

Reviewer: ```suggestion

== What's next?
```

> 
> To learn more about using the Pulsar Admin Console, see xref:admin-console-tutorial.adoc[Admin Console Tutorial].
> To learn more about installing Pulsar on a server or virtual machine, see xref:quickstart-server-installs.adoc[Quickstart for Server/VM Install].

Reviewer: ```suggestion
Once your pods are up and running, continue to xref:quickstart-helm-installs.adoc#manage-pulsar-cluster[Accessing Pulsar Cluster in Cloud] to access and manage your cluster. 

* To learn more about using the Pulsar Admin Console, see xref:admin-console-tutorial.adoc[Admin Console Tutorial].
* To learn more about installing Pulsar on a server or virtual machine, see xref:quickstart-server-installs.adoc[Quickstart for Server/VM Install].
```

For more on Helm, see https://helm.sh/docs/[Helm documentation].
For more on Minikube, see https://minikube.sigs.k8s.io/docs/start/[Minikube documentation]. 

== Provision Your GKE Cluster

Before you can install the Helm chart, you need to configure the storage class settings for your cloud provider. The handling of storage varies from cloud provider to cloud provider.

Create a new file called `storage_values.yaml` for the storage class settings. This allows you to match the optimum storage type to the volume.
If you have specific storage class requirement, for example fixed IOPS disks in AWS, you can have the chart configure the storage classes for you.
Here is an example `storage-values.yaml` for GKE.

----
# For GKE
# default_storage:
#   provisioner: kubernetes.io/gce-pd
#   type: pd-ssd
#   fsType: ext4
#   extraParams:
#      replication-type: none
----

See this https://github.com/datastax/pulsar-helm-chart/blob/master/helm-chart-sources/pulsar/values.yaml[values file] for more details on the settings in `values.yaml`.

Additional provisioning set-up options are available https://docs.k8ssandra.io/install/gke/[here].

== Install Luna Streaming to GKE

> 

Reviewer: @mendonk Please make these changes or resolve if you already have. Thanks!

Once you have added your storage settings to the `storage-values.yaml` file, install the Helm chart. 

> . Create the namespace; in this example, we use `pulsar`.

Reviewer 2: Don't think this is needed

Reviewer: @mendonk Please make these changes or resolve if you already have. Thanks!

+
----
kubectl create namespace pulsar
----

. Run this Helm command using the `storage-values.yaml` provisioning:
+
----
helm install pulsar datastax-pulsar/pulsar --namespace pulsar --values storage_values.yaml --create-namespace
----

. To avoid having to specify the `pulsar` namespace on each subsequent command, set the namespace context. Example:
+
----
kubectl config set-context $(kubectl config current-context) --namespace=pulsar
----

. Execute `kubectl get pods -A` to view your running pods. 
. Once your pods are up and running, continue to xref:quickstart-helm-installs.adoc#manage-pulsar-cluster[Accessing Pulsar Cluster in Cloud] to access and manage your cluster. 

> == Next

Reviewer 2: As a general comment, especially after I also took a look at the other 2 cloud provider specific documents, I don' think these documents are really necessary beyond the existing Helm Chart doc: https://docs.datastax.com/en/luna/streaming/2.8/quickstart-helm-installs.html

The only difference in these documents is about the storage class difference which is already covered by the Helm Chart doc. Other than that, there is nothing cloud provider specific stuff. 

So unless we want to describe the procedure differences about how to launch the K8s clusters in these different cloud providers, I don't think there are much value added by having these 3 documents.

Reviewer: @mendonk Please make these changes or resolve if you already have. Thanks!
